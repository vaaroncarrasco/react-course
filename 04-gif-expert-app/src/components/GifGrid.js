import React from 'react'
import PropTypes from 'prop-types';
import { useFetchGifs } from '../hooks/useFetchGifs'
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({ category }) => {

    // Every time theres a change in component, it runs the whole code all over again and again -> useEffect fixes this by calling functions once

    const { data:images, loading } = useFetchGifs( category );

    return (
        <>
            <h3> { category } </h3>

            { loading && <p>Loading...</p> }

            <div className="card-grid">
                {
                    images.map(img => (
                        <GifGridItem
                            key={ img.id }
                            // * spreading img{} props and sending em all at once instead of prop={ prop }
                            { ...img }
                            // img={ img }
                        />
                    ))
                }
            </div>
        </>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}