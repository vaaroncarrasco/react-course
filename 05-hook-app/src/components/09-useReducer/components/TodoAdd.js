import React from 'react'
import { useForm } from '../../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo }) => {

    const [ { description }, handleInputChange, reset ] = useForm({ // useForm returns [values{}, handleInputChange]
        // same as name="description" in html attributes
        description: '',
    });

    const handleSubmit = (e) => {
        e.preventDefault();

        if ( description.trim().length <= 1) return;

        // * new object
        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        }

        // * dispatch sends action to reducer
        handleAddTodo( newTodo );
        reset();

    }

    return (
        <>
            <h4>Agregar ToDo</h4>
            <hr />

            <form onSubmit={ handleSubmit }>
                <input
                    type="text"
                    name="description"
                    className="form-control"
                    placeholder="Aprender..."
                    autoComplete="off"
                    value={ description }
                    onChange={ handleInputChange }
                />

                <button
                    type="submit"
                    className="btn btn-outline-primary mt-1 btn-block"
                >
                    Agregar
                </button>

            </form>
        </>
    )
}
