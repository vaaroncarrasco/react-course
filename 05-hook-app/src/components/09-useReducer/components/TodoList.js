import React from 'react'
import { TodoListItem } from './TodoListItem'
import PropTypes from 'prop-types';

export const TodoList = ({ todos, handleDelete, handleToggle }) => {
    return (
        <ul className="list-group list-group-flush">
            {
                todos.map( (todo, index) => ( // * () => () -> () implicitly returning an object
                    <TodoListItem
                        key={ todo.id }
                        todo={ todo }
                        index={ index }
                        handleDelete={ handleDelete }
                        handleToggle={ handleToggle }
                    />
                ))
            }
        </ul>
    )
}

TodoList.propTypes = {
    todos: PropTypes.array.isRequired,
    handleDelete: PropTypes.func.isRequired,
    handleToggle: PropTypes.func.isRequired
}
