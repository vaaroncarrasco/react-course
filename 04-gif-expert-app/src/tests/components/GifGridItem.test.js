import React from 'react';
import '@testing-library/jest-dom'
import { shallow } from 'enzyme';
import { GifGridItem } from '../../components/GifGridItem';

describe('GifGridItem testing', () => {

    const title = 'hola mundo';
    const url = 'https://localhost/hola.gif';
    const wrapper = shallow(<GifGridItem title={ title } url={ url }/>);

    test('should show component and match snapshot', () => {
        expect( wrapper ).toMatchSnapshot();
    });

    test('should have a p with title', () => {
        const p = wrapper.find('p');
        expect( p.text().trim() ).toBe( title );
    });

    test('should have an img with src=url & alt=title', () => {
        const img = wrapper.find('img');

        expect( img.prop('src') ).toBe( url );
        expect( img.prop('alt') ).toBe( title );
    });

    test('should have animate_fadeIn', () => {
        const div = wrapper.find('div');
        const className = div.prop('className');
        expect( className ).toBe( 'card animate__animated animate__fadeIn' );
    });

});
