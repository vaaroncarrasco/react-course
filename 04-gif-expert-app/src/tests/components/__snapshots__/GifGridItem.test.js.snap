// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GifGridItem testing should show component and match snapshot 1`] = `
<div
  className="card animate__animated animate__fadeIn"
>
  <img
    alt="hola mundo"
    src="https://localhost/hola.gif"
  />
  <p>
     
    hola mundo
     
  </p>
</div>
`;
